transaction/incoming.go

```diff
@@ -0,0 +1,137 @@
+package transaction
+
+import (
+	"math"
+	"time"
+
+	"github.com/levenlabs/go-llog"
+	"github.com/levenlabs/saturn/config"
+	lproto "github.com/levenlabs/saturn/proto"
+)
+
+// IncomingMessage takes a message just received from the remote address and
+// processes it. It may return a message which should be sent back to the
+// remote, or it may return nil
+func IncomingMessage(msg *lproto.TxMsg) *lproto.TxMsg {
+	im := incomingMsg{
+		msg:     msg,
+		replyCh: make(chan *lproto.TxMsg),
+	}
+	incomingMsgCh <- im
+	return <-im.replyCh
+}
+
+type incomingMsg struct {
+	msg     *lproto.TxMsg
+	replyCh chan *lproto.TxMsg
+}
+
+var incomingMsgCh = make(chan incomingMsg)
+
+func incoming(transactions map[string]*tx, msg *lproto.TxMsg) *lproto.TxMsg {
+
+	kv := llog.KV{
+		"msg": msg,
+	}
+	if !msg.Verify() {
+		llog.Warn("received incorrectly signed message", kv)
+		return nil
+	}
+
+	t := transactions[msg.Id]
+	if t != nil && t.expectedSeq != msg.Seq {
+		return nil
+	}
+
+	retTx := &lproto.TxMsg{
+		Id:  msg.Id,
+		Seq: msg.Seq + 1,
+	}
+
+	var ok bool
+	if r := msg.GetInitialReport(); r != nil {
+		retTx.Inner, ok = handleIncomingInitialReport(transactions, msg.Id, r)
```

Comment:
Don't pass in `msg.Id` here since this should be generating that. Unless we want to support the slave making an `id` and only generating one on the master when the slave DOES NOT send one?